crates/networking/rpc/engine/fork_choice.rs

```diff
@@ -220,17 +220,39 @@ fn handle_forkchoice(
                 }
             };
 
-            // Check if the block has already been invalidated
-            match invalid_ancestors.get(&fork_choice_state.head_block_hash) {
-                Some(latest_valid_hash) => {
-                    Err(InvalidForkChoice::InvalidAncestor(*latest_valid_hash))
+            // Check head block hash in invalid_ancestors
+            if let Some(latest_valid_hash) =
+                invalid_ancestors.get(&fork_choice_state.head_block_hash)
+            {
+                warn!(
+                    "Invalid fork choice state. Reason: Invalid ancestor {:#x}",
+                    latest_valid_hash
+                );
+                Err(InvalidForkChoice::InvalidAncestor(*latest_valid_hash))
+            } else {
```

Comment:
nit, I would avoid the else and do early returns so that is more clearer, for example:

```suggestion
            if let Some(latest_valid_hash) =
                invalid_ancestors.get(&fork_choice_state.head_block_hash)
            {
                warn!(
                    "Invalid fork choice state. Reason: Invalid ancestor {:#x}",
                    latest_valid_hash
                );
               return Err(InvalidForkChoice::InvalidAncestor(*latest_valid_hash));
            }
            
            let check_parent = ....
```

Reply:
I agree with the idea! Some of the changes i made in the new_Payload handles in this PR do exactly this. The issue here is that it's not possible to return early because we are inside a 
  `let fork_choice_res = match context.sync_status()? {`

If we want it to be an early return we'll need to extract this logic as new functions and refactor the rest of the `handle_forkchoice` which I agree it would be better for readability but probably something that could be done in another PR. But if others think the same I could do it here!